<a href="https://colab.research.google.com/github/THESHAMIK/Real-Life-Projects/blob/main/Deep_NN/AutoKeras_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Here we tried to re-visit AUTOKERAS & tried to optimize it using more epochs, max trails, loss, metrics, callbacks, objective, tuner etc as per below site:
https://autokeras.com/structured_data_regressor/

Done most of experiments with both X & y as SCALED.

https://machinelearningmastery.com/autokeras-for-classification-and-regression/ 



In [ ]:
# Import packages
import numpy as np
import pandas as pd
import random
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from keras.models import Sequential
from keras.layers import Dense, BatchNormalization, Dropout
#from keras.optimizers import SGD, RMSprop, Adadelta, Adagrad, Adamax, Nadam, Ftrl
from tensorflow.keras.optimizers import SGD, RMSprop, Adadelta, Adagrad, Adamax, Nadam, Ftrl
from tensorflow.keras.optimizers import Adam ##because cannot import name 'Adam' from 'keras.optimizers'
from keras.callbacks import EarlyStopping, ModelCheckpoint
#from keras.wrappers.scikit_learn import KerasClassifier
from keras.wrappers.scikit_learn import KerasRegressor
from math import floor
from sklearn.metrics import make_scorer, accuracy_score
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import r2_score
#from bayes_opt import BayesianOptimization
from sklearn.model_selection import StratifiedKFold
from keras.layers import LeakyReLU
LeakyReLU = LeakyReLU(alpha=0.1)
import tensorflow as tf

import warnings
warnings.filterwarnings('ignore')
pd.set_option("display.max_columns", None)

In [ ]:
df = pd.read_csv("khds_final_1.csv")
df['Product Sulphur'] = pd.to_numeric(df['Product Sulphur'],errors = 'coerce') ##==> This one works as pd.to_numeric works for convert a list, a series, an array, or a tuple to a numeric datatype
filtered_df = df[df['Product Sulphur'].notnull()]
df.dropna(axis=0)
filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 711 entries, 0 to 710
Data columns (total 34 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   LABfeed Sulphur                            711 non-null    float64
 1   T1 : FRESH FEED FLOW TO KHDS KGPerHour     711 non-null    float64
 2   T1 : Reactor inlet temperature DEGC        711 non-null    float64
 3   T1 : Hydrogen make-up flow NM3perHour      711 non-null    float64
 4   T1 : Recycle hydrogen flow NM3perHour      711 non-null    float64
 5   T1 : REACTOR OUTLET Temp DEGC              711 non-null    float64
 6   T1 : REACTOR FEED INLET Pressure KGperCM2  711 non-null    float64
 7   T1 : Cold separator pressure KGperCM2      711 non-null    float64
 8   T1 : KERO FEED DENSITY KGperCM2            711 non-null    float64
 9   T2 : FRESH FEED FLOW TO KHDS KGPerHour     711 non-null    float64
 10  T2 : Reactor inlet tempera

In [ ]:
target_col = "Product Sulphur"
X = filtered_df.loc[:, filtered_df.columns != target_col]
y = filtered_df.loc[:, target_col]

In [ ]:
from sklearn import metrics
import pandas
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasRegressor
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from keras.activations import linear
import tensorflow.keras as keras
from tensorflow.keras import regularizers
import matplotlib.pyplot as plt
from keras.models import Sequential
from tensorflow.keras.optimizers import Adam# define model
from keras.optimizer_v2.adam import Adam

In [ ]:
y = y.to_numpy()
y = y.reshape(-1,1)
#trainy = trainy.to_numpy()
#trainy = trainy.reshape(-1,1)

In [ ]:
#from sklearn.preprocessing import StandardScaler
#import numpy as np
# created scaler for only X predictors
input_scaler = StandardScaler()
# fit scaler on training dataset
# fit scaler
input_scaler.fit(y)
# transform training dataset
y_scaled = input_scaler.transform(y)
# transform test dataset
#testX_scaled = input_scaler.transform(testX)

In [ ]:
#from sklearn.preprocessing import StandardScaler
#import numpy as np
# created scaler for only X predictors
input_scaler = StandardScaler()
# fit scaler on training dataset
# fit scaler
input_scaler.fit(X)
# transform training dataset
X_scaled = input_scaler.transform(X)
# transform test dataset
#testX_scaled = input_scaler.transform(testX)

In [ ]:
y.shape

(711, 1)

In [ ]:
!pip install autokeras

     |████████████████████████████████| 162 kB 6.8 MB/s 
     |████████████████████████████████| 133 kB 70.2 MB/s 
     |████████████████████████████████| 462 kB 65.2 MB/s 


In [ ]:
import plotly.express as px
import autokeras as ak
from autokeras import StructuredDataRegressor
from sklearn.model_selection import train_test_split
from IPython.display import HTML

In [ ]:
earlystopper = EarlyStopping(monitor='loss', min_delta=0, patience=15, verbose=1, mode='auto') #patience reduced to 10 from 15
#estimator = KerasRegressor(build_fn=baseline_model, epochs=100, batch_size=5, verbose=0, shuffle=False)
#history_5_4 = lr_model_d.fit(trainX_scaled, trainy_scaled, batch_size=30, epochs = 500, validation_data=(testX_scaled, testy_scaled),shuffle = True, verbose = 0, callbacks = [earlystopper])
#history_5_4 = lr_model_d_1.fit(trainX, trainy_scaled, batch_size=30, epochs = 500, validation_data=(testX, testy_scaled),shuffle = True, verbose = 0, callbacks = [earlystopper])
#estimator = KerasRegressor(build_fn=baseline_model, epochs=100, batch_size=5, verbose=0, shuffle=False,callbacks=[EarlyStopping(monitor='val_loss', patience=20)])
#estimator = KerasRegressor(build_fn=baseline_model, epochs=100, batch_size=5, verbose=0, shuffle=False,callbacks=[earlystopper])

In [ ]:
# It tries 10 different models.
reg = ak.StructuredDataRegressor(max_trials=40, overwrite=True, loss="mean_squared_error",metrics=None, objective="val_loss")
#reg = ak.StructuredDataRegressor(overwrite=True, loss="mean_squared_error",metrics=None, objective="val_loss", tuner='random')
# Feed the structured data regressor with training data.
reg.fit(X_scaled, y_scaled, epochs=200, callbacks=[earlystopper])

Trial 40 Complete [00h 00m 09s]
val_loss: 4.023104667663574

Best val_loss So Far: 2.275585651397705
Total elapsed time: 00h 19m 59s
INFO:tensorflow:Oracle triggered exit
Epoch 1/200
23/23 [==============================] - 1s 4ms/step - loss: 1.2016 - mean_squared_error: 1.2016
Epoch 2/200
23/23 [==============================] - 0s 3ms/step - loss: 1.6616 - mean_squared_error: 1.6616
Epoch 3/200
23/23 [==============================] - 0s 3ms/step - loss: 0.9121 - mean_squared_error: 0.9121
Epoch 4/200
23/23 [==============================] - 0s 4ms/step - loss: 0.8055 - mean_squared_error: 0.8055
Epoch 5/200
23/23 [==============================] - 0s 4ms/step - loss: 0.7894 - mean_squared_error: 0.7894
Epoch 6/200
23/23 [==============================] - 0s 4ms/step - loss: 0.7485 - mean_squared_error: 0.7485
Epoch 7/200
23/23 [==============================] - 0s 4ms/step - loss: 0.7397 - mean_squared_error: 0.7397
Epoch 8/200
23/23 [==============================] - 0s 3ms/step -

Trial 3 Complete [00h 00m 03s]
val_loss: 4.315784931182861

Best val_loss So Far: 3.6199324131011963
Total elapsed time: 00h 00m 33s
Epoch 100/100
23/23 [==============================] - 0s 3ms/step - loss: 0.2593 - mean_squared_error: 0.2593
reg = ak.StructuredDataRegressor(max_trials=3, overwrite=True) ==> These were params on first pass1.

Pass2 ::
Trial 3 Complete [00h 00m 21s]
val_loss: 2.3486874103546143

Best val_loss So Far: **2.3**486874103546143
Total elapsed time: 00h 00m 44s

reg = ak.StructuredDataRegressor(max_trials=3, overwrite=True, loss="mean_squared_error",metrics=None, objective="val_loss")

Pass 3: With callbacks & max_trails = 30 from 3::
Trial 30 Complete [00h 00m 29s]
val_loss: 2.5138509273529053

Best val_loss So Far: **2.303**161382675171
Total elapsed time: 00h 10m 55s

PASS 4: tuner='bayesian' added but max_trails = 15 from 25 (hyperband was worst)::
Trial 25 Complete [00h 00m 05s]
val_loss: 4.327514171600342

Best val_loss So Far: 2.7460367679595947

PASS 5: tuner ='random' max trail = 25
Trial 25 Complete [00h 00m 05s]
val_loss: 4.380385398864746

Best val_loss So Far: 2.49894642829895
Total elapsed time: 00h 07m 32s

SO far randim works best among all tuners = bayesian, hyperband, greedy.

pass6: same as pass5 except max trails removed. by default its 100 but why callback not owkring?
Trial 100 Complete [00h 00m 16s]
val_loss: 4.087159156799316

Best val_loss So Far: 2.495527505874634
Total elapsed time: 00h 26m 50s

pass 7: all same above pass6, except epoch =200 from 100 & max trail =40
Trial 40 Complete [00h 00m 10s]
val_loss: 4.36867094039917

Best val_loss So Far: 2.4152731895446777
Total elapsed time: 00h 14m 13s

pass 8: same as pass 7 witrh epoch =300 from 200
Trial 40 Complete [00h 00m 06s]
val_loss: 4.809711933135986

Best val_loss So Far: 3.2346599102020264
Total elapsed time: 00h 10m 55s

PASS 9:: Replicating pass 3 with max trails = 40, epochs =200 & no tuner used
Trial 40 Complete [00h 00m 09s]
val_loss: 4.023104667663574

Best val_loss So Far: **2.275**585651397705
Total elapsed time: 00h 19m 59s

reg = ak.StructuredDataRegressor(max_trials=40, overwrite=True, loss="mean_squared_error",metrics=None, objective="val_loss")

pass 11: All same as pass 10 but split rati set to 0.1 than 0.2 above. Also earlystopper = EarlyStopping(monitor='mean_squared_error', min_delta=0, patience=15, verbose=1, mode='auto')--> monitor changed from 'loss' to mse

Trial 30 Complete [00h 00m 32s]
val_loss: 0.21225327253341675

Best val_loss So Far: **0.121**28239125013351



In [ ]:
# Predict with the best model.
predicted_y_scaled = reg.predict(X_scaled)

23/23 [==============================] - 0s 5ms/step


In [ ]:
from sklearn import metrics
score = np.sqrt(metrics.mean_squared_error(predicted_y_scaled,y_scaled))
print(f"Before save score (RMSE): {score}")
# Measure MSE error.  
score = metrics.mean_squared_error(predicted_y_scaled,y_scaled)
print("Final score (MSE): {}".format(score))
score1 = metrics.mean_absolute_error(predicted_y_scaled,y_scaled)
print("Final score (MAE): {}".format(score1))
score2 = metrics.r2_score(predicted_y_scaled,y_scaled)
print("Final score (R2): {}".format(score2))
score3 = metrics.mean_absolute_percentage_error(predicted_y_scaled,y_scaled)
print("Final score (MAPE): {}".format(score3))
#score4 = metrics.mean_squared_log_error(predicted_y_scaled,testy_scaled)
#print("Final score (MSLE): {}".format(score4))
#score5 = metrics.log_loss(predicted_y_scaled,testy_scaled)

Before save score (RMSE): 0.33888006070618804
Final score (MSE): 0.1148396955442297
Final score (MAE): 0.20871200024821696
Final score (R2): 0.8726204283480066
Final score (MAPE): 3.9401550287474296


Earlier even with both scaling result was (for all params being same):
Final score (MSE): 6.122893377762371
Final score (MAE): 0.8093436874593429
Final score (R2): -17.210139256045256
Final score (MAPE): 5.452198501882145

Now on same params result is ::
Before save score (RMSE): 0.5015663350450803
Final score (MSE): 0.25156878845055375
Final score (MAE): 0.2706559128012366
Final score (R2): **0.608**6809130830875
Final score (MAPE): 6.073575709195961

All params same yes, BUT only diff is using X(scaled),y(scaled) in totality without splitting. So results were good as model got more data for training itself but result good
Before save score (RMSE): 0.3977670902847305
Final score (MSE): 0.15821865811358096
Final score (MAE): 0.23472967040490217
Final score (R2): **0.81**51491010882141
Final score (MAPE): 3.8867089743956265

PASS3:  With callbacks & max_trails = 30 from 3 (but still with full X & no splitting)::

Before save score (RMSE): 0.3536620359655597
Final score (MSE): 0.12507683568330483
Final score (MAE): 0.23672821980790304
Final score (R2): **0.8761**266873177348
Final score (MAPE): 8.117779046966902

PASS4: Added tuner = 'bayesian' (hyper band worked very bad). Max trail = 25

Before save score (RMSE): 0.3860245383558159
Final score (MSE): 0.14901494421282077
Final score (MAE): 0.24365072890837666
Final score (R2): 0.7921041448849359
Final score (MAPE): 3.3612951371158633

pass 5: Added tuner = 'random' (hyper band worked very bad). Max trail = 25

Before save score (RMSE): 0.3779711353591584
Final score (MSE): 0.1428621791646912
Final score (MAE): 0.19481394413346476
Final score (R2): 0.8259287165772973
Final score (MAPE): 10.418282811744374

PASS 6:Same as pass 5 except max trails removed so its equal to 100
Before save score (RMSE): 0.34210636155473045
Final score (MSE): 0.11703676261621594
Final score (MAE): 0.18231880587676905
Final score (R2): 0.8588170049866046
Final score (MAPE): 35.44145571959353
No much different from max trails = 25

pass 7:all same as pass6 exept epoch=200 & max trail 40 from 100
Before save score (RMSE): 0.3545376196520452
Final score (MSE): 0.12569692374853828
Final score (MAE): 0.18252120878757355
Final score (R2): 0.8411879757138232
Final score (MAPE): 5.190196233708742

PASS 9:: Replicating pass 3 with max trails = 40, epochs =200 & no tuner used
Before save score (RMSE): 0.33888006070618804
Final score (MSE): 0.1148396955442297
Final score (MAE): 0.20871200024821696
Final score (R2): **0.872**6204283480066
Final score (MAPE): 3.9401550287474296

reg = ak.StructuredDataRegressor(max_trials=40, overwrite=True, loss="mean_squared_error",metrics=None, objective="val_loss")

PASS 10: Look at huge bad scores when train/test split is used:: All other criteria is same as above pass 9 that is epoch=200, max trail=40, no tuner used.

Before save score (RMSE): 0.8593099630341421
Final score (MSE): 0.7384136125697386
Final score (MAE): 0.3347192538045948
Final score (R2): -1.1029342767960197
Final score (MAPE): 7.819765244431312

pass 11: All same as pass 10 but split rati set to 0.1 than 0.2 above. Also earlystopper = EarlyStopping(monitor='mean_squared_error', min_delta=0, patience=15, verbose=1, mode='auto')--> monitor changed from 'loss' to mse

Trial 30 Complete [00h 00m 32s]
val_loss: 0.21225327253341675

Best val_loss So Far: **0.121**28239125013351
Before save score (RMSE): 0.3482562101307629
Final score (MSE): 0.12128238789464207
Final score (MAE): 0.26273287179509514
Final score (R2): 0.6448275219970139
Final score (MAPE): 2.2944849171203074

pass 12:::same as pass 10 but monitor = mse than loss. scores same as pass 10
Trial 40 Complete [00h 00m 20s]
val_loss: 0.7478248476982117

Best val_loss So Far: 0.7034339308738708
Total elapsed time: 00h 11m 36s
Before save score (RMSE): 0.8387097153018569
Final score (MSE): 0.7034339865417218
Final score (MAE): 0.328256171613517
Final score (R2): -0.9619806941781535
Final score (MAPE): 4.187767267610768


In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y_scaled, test_size=0.20, random_state=2021)

**Difference between validation split & validation data**
validation_split: The best model found would be fit on the entire dataset **including the validation data**.

validation_data:  The best model found would be fit on the training dataset **without the validation** data.

In [ ]:
earlystopper = EarlyStopping(monitor='mean_squared_error', min_delta=0, patience=15, verbose=1, mode='auto') #patience reduced to 10 from 15
reg = ak.StructuredDataRegressor(max_trials=40, overwrite=True, loss="mean_squared_error",metrics=None, objective="val_loss")
reg.fit(X_train, y_train, validation_data= (X_test, y_test), epochs=200, callbacks=[earlystopper])

Trial 40 Complete [00h 00m 20s]
val_loss: 0.7478248476982117

Best val_loss So Far: 0.7034339308738708
Total elapsed time: 00h 11m 36s
INFO:tensorflow:Oracle triggered exit
INFO:tensorflow:Assets written to: ./structured_data_regressor/best_model/assets


In [ ]:
# Predict with the best model.
predicted_y = reg.predict(X_test)

5/5 [==============================] - 0s 3ms/step


In [ ]:
from sklearn import metrics
score = np.sqrt(metrics.mean_squared_error(predicted_y,y_test))
print(f"Before save score (RMSE): {score}")
# Measure MSE error.  
score = metrics.mean_squared_error(predicted_y,y_test)
print("Final score (MSE): {}".format(score))
score1 = metrics.mean_absolute_error(predicted_y,y_test)
print("Final score (MAE): {}".format(score1))
#score2 = metrics.r2_score(predicted_y,y_test)
score2 = metrics.r2_score(y_test, predicted_y)
print("Final score (R2): {}".format(score2))
score3 = metrics.mean_absolute_percentage_error(predicted_y,y_test)
print("Final score (MAPE): {}".format(score3))
#score4 = metrics.mean_squared_log_error(predicted_y_scaled,testy_scaled)
#print("Final score (MSLE): {}".format(score4))
#score5 = metrics.log_loss(predicted_y_scaled,testy_scaled)

Before save score (RMSE): 0.8387097153018569
Final score (MSE): 0.7034339865417218
Final score (MAE): 0.328256171613517
Final score (R2): 0.3096852143066814
Final score (MAPE): 4.187767267610768


Look at huge reduction of R2 score when train/test split is used though Scaled COMPARED to when no spilitting is used
Before save score (RMSE): 0.8593099630341421
Final score (MSE): 0.7384136125697386
Final score (MAE): 0.3347192538045948
Final score (R2): -1.1029342767960197
Final score (MAPE): 7.819765244431312



In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y_scaled, test_size=0.10, random_state=2021)

In [ ]:
earlystopper = EarlyStopping(monitor='mean_squared_error', min_delta=0, patience=15, verbose=1, mode='auto') #patience reduced to 10 from 15
reg_1 = ak.StructuredDataRegressor(max_trials=40, overwrite=True, loss="mean_squared_error",metrics=None, objective="val_loss")
reg_1.fit(X_train, y_train, validation_data= (X_test, y_test), epochs=200, callbacks=[earlystopper])

Trial 40 Complete [00h 00m 17s]
val_loss: 0.18754418194293976

Best val_loss So Far: 0.1450742930173874
Total elapsed time: 00h 11m 34s
INFO:tensorflow:Oracle triggered exit
INFO:tensorflow:Assets written to: ./structured_data_regressor/best_model/assets


In [ ]:
# Predict with the best model.
predicted_y = reg_1.predict(X_test)

3/3 [==============================] - 0s 4ms/step


In [ ]:
from sklearn import metrics
score = np.sqrt(metrics.mean_squared_error(predicted_y,y_test))
print(f"Before save score (RMSE): {score}")
# Measure MSE error.  
score = metrics.mean_squared_error(predicted_y,y_test)
print("Final score (MSE): {}".format(score))
score1 = metrics.mean_absolute_error(predicted_y,y_test)
print("Final score (MAE): {}".format(score1))
score2 = metrics.r2_score(predicted_y,y_test)
print("Final score (R2): {}".format(score2))
score3 = metrics.mean_absolute_percentage_error(predicted_y,y_test)
print("Final score (MAPE): {}".format(score3))
#score4 = metrics.mean_squared_log_error(predicted_y_scaled,testy_scaled)
#print("Final score (MSLE): {}".format(score4))
#score5 = metrics.log_loss(predicted_y_scaled,testy_scaled)

Before save score (RMSE): 0.38088619742505564
Final score (MSE): 0.14507429538891845
Final score (MAE): 0.2585138127133954
Final score (R2): 0.43282060707956005
Final score (MAPE): 6.891545857502718


Before save score (RMSE): 0.3482562101307629
Final score (MSE): 0.12128238789464207
Final score (MAE): 0.26273287179509514
Final score (R2): 0.6448275219970139
Final score (MAPE): 2.2944849171203074

Before save score (RMSE): 0.3979641240994279
Final score (MSE): 0.15837544407022483
Final score (MAE): 0.24911958268606466
Final score (R2): 0.4401897757193526
Final score (MAPE): 71.78623126083743

PASS 3:
Trial 40 Complete [00h 00m 17s]
val_loss: 0.18754418194293976

Best val_loss So Far: 0.1450742930173874

Before save score (RMSE): 0.38088619742505564
Final score (MSE): 0.14507429538891845
Final score (MAE): 0.2585138127133954
Final score (R2): 0.43282060707956005
Final score (MAPE): 6.891545857502718

In [ ]:
auto_keras_model = reg_1.export_model()
print(type(auto_keras_model)) 

<class 'keras.engine.functional.Functional'>


In [ ]:
try:
    auto_keras_model.save("model_autokeras", save_format="tf")
except Exception:
    auto_keras_model.save("model_autokeras.h5")

INFO:tensorflow:Assets written to: model_autokeras/assets


In [ ]:
from tensorflow.keras.models import load_model

In [ ]:
from tensorflow.keras.models import load_model
loaded_model = load_model("model_autokeras", custom_objects=ak.CUSTOM_OBJECTS)

After saving & loading model with split = 0.1 & r2=0.64 - pass 10
Before save score (RMSE): 100036.98826900923
Final score (MSE): 10007399021.933891
Final score (MAE): 99759.53966346153
Final score (R2): -179.5118581603864
Final score (MAPE): 0.9999725206664197
But this was x unseen & y both UNSCALED. Lets scale them:

pass2:
Before save score (RMSE): 4.4050792542843915
Final score (MSE): 19.404723236526728
Final score (MAE): 2.527393555993899
Final score (R2): -0.4088693349926624
Final score (MAPE): 3.684403254055156
NOT GOOD RESULTS

In [ ]:
df_test = pd.read_csv("KHDS_TestData_12Apr_rev_1.csv")
'''df_test['Product Sulphur'] = pd.to_numeric(df['Product Sulphur'],errors = 'coerce') ##==> This one works as pd.to_numeric works for convert a list, a series, an array, or a tuple to a numeric datatype
df_test['LABfeed Sulphur'] = pd.to_numeric(df['LABfeed Sulphur'],errors = 'coerce')
filtered_df_test = df_test[df_test['Product Sulphur'].notnull()]
filtered_df_test = df_test[df_test['LABfeed Sulphur'].notnull()]
df.dropna(axis=0)'''
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 34 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   LABfeed Sulphur                            52 non-null     int64  
 1   T1 : FRESH FEED FLOW TO KHDS KGPerHour     52 non-null     float64
 2   T1 : Reactor inlet temperature DEGC        52 non-null     float64
 3   T1 : Hydrogen make-up flow NM3perHour      52 non-null     float64
 4   T1 : Recycle hydrogen flow NM3perHour      52 non-null     float64
 5   T1 : REACTOR OUTLET Temp DEGC              52 non-null     float64
 6   T1 : REACTOR FEED INLET Pressure KGperCM2  52 non-null     float64
 7   T1 : Cold separator pressure KGperCM2      52 non-null     float64
 8   T1 : KERO FEED DENSITY KGperCM2            52 non-null     float64
 9   T2 : FRESH FEED FLOW TO KHDS KGPerHour     52 non-null     float64
 10  T2 : Reactor inlet temperatu

In [ ]:
target_col = "Product Sulphur"
X_unseen = df_test.loc[:, df_test.columns != target_col]
y_unseen = df_test.loc[:, target_col]

In [ ]:
X_unseen.shape

(52, 33)

In [ ]:
y_unseen.shape

(52,)

In [ ]:
y_unseen

0      44
1      42
2      39
3       6
4       6
5       7
6       6
7       4
8       3
9       3
10      3
11      3
12      3
13      5
14      4
15      5
16      4
17      4
18      4
19     60
20     90
21     40
22     45
23    135
24     80
25    184
26     11
27      6
28      7
29      4
30      7
31      5
32      8
33      5
34      7
35      4
36      7
37      4
38      3
39      5
40      6
41      4
42      4
43      4
44     10
45      5
46      5
47      2
48      3
49     87
50     80
51     75
Name: Product Sulphur, dtype: int64

In [ ]:
#from sklearn.preprocessing import StandardScaler
#import numpy as np
# created scaler for only X predictors
input_scaler = StandardScaler()
# fit scaler on training dataset
# fit scaler
input_scaler.fit(X_unseen)
# transform training dataset
X_unseen_scaled = input_scaler.transform(X_unseen)
# transform test dataset
#testX_scaled = input_scaler.transform(testX)

In [ ]:
y_unseen = y_unseen.to_numpy()
y_unseen = y_unseen.reshape(-1,1)
#trainy = trainy.to_numpy()
#trainy = trainy.reshape(-1,1)

In [ ]:
#from sklearn.preprocessing import StandardScaler
#import numpy as np
# created scaler for only X predictors
input_scaler = StandardScaler()
# fit scaler on training dataset
# fit scaler
input_scaler.fit(y_unseen)
# transform training dataset
y_unseen_scaled = input_scaler.transform(y_unseen)
# transform test dataset
#testX_scaled = input_scaler.transform(testX)

In [ ]:
from tensorflow.keras.models import load_model
loaded_model = load_model("model_autokeras", custom_objects=ak.CUSTOM_OBJECTS)

In [ ]:
predicted_y = loaded_model.predict(X_unseen_scaled)

In [ ]:
from sklearn import metrics
score = np.sqrt(metrics.mean_squared_error(predicted_y,y_unseen_scaled))
print(f"Before save score (RMSE): {score}")
# Measure MSE error.  
score = metrics.mean_squared_error(predicted_y,y_unseen_scaled)
print("Final score (MSE): {}".format(score))
score1 = metrics.mean_absolute_error(predicted_y,y_unseen_scaled)
print("Final score (MAE): {}".format(score1))
score2 = metrics.r2_score(predicted_y,y_unseen_scaled)
print("Final score (R2): {}".format(score2))
score3 = metrics.mean_absolute_percentage_error(predicted_y,y_unseen_scaled)
print("Final score (MAPE): {}".format(score3))
#score4 = metrics.mean_squared_log_error(predicted_y_scaled,testy_scaled)
#print("Final score (MSLE): {}".format(score4))
#score5 = metrics.log_loss(predicted_y_scaled,testy_scaled)

Before save score (RMSE): 4.601016195809429
Final score (MSE): 21.169350034100667
Final score (MAE): 2.8865882105606144
Final score (R2): -0.40311635298566184
Final score (MAPE): 1.3468413555834309


first pass :
Before save score (RMSE): 4.4050792542843915
Final score (MSE): 19.404723236526728
Final score (MAE): 2.527393555993899
Final score (R2): -0.4088693349926624
Final score (MAPE): 3.684403254055156

2nd pass:

Before save score (RMSE): 2.133150516600042
Final score (MSE): 4.5503311264710256
Final score (MAE): 1.2227432087982566
Final score (R2): **0.249**5769624008798
Final score (MAPE): 2.7604413954128613

3RD PASS:
Before save score (RMSE): 1.6672130394726354
Final score (MSE): 2.779599318987583
Final score (MAE): 0.8111153610804643
Final score (R2): **0.57**76599931787793
Final score (MAPE): 0.4814028452057367

In [ ]:
type(predicted_y)

numpy.ndarray

In [ ]:
type(y_unseen_scaled)

numpy.ndarray

In [ ]:
type(X_unseen_scaled)

numpy.ndarray

In [ ]:
predicted_y.shape

(52, 1)

In [ ]:
y_unseen_scaled.shape

(52, 1)

In [ ]:
y_unseen_scaled

array([[ 0.56307932],
       [ 0.50940356],
       [ 0.42888993],
       [-0.45676003],
       [-0.45676003],
       [-0.42992216],
       [-0.45676003],
       [-0.51043579],
       [-0.53727367],
       [-0.53727367],
       [-0.53727367],
       [-0.53727367],
       [-0.53727367],
       [-0.48359791],
       [-0.51043579],
       [-0.48359791],
       [-0.51043579],
       [-0.51043579],
       [-0.51043579],
       [ 0.99248536],
       [ 1.79762169],
       [ 0.45572781],
       [ 0.5899172 ],
       [ 3.00532619],
       [ 1.52924292],
       [ 4.32038219],
       [-0.32257065],
       [-0.45676003],
       [-0.42992216],
       [-0.51043579],
       [-0.42992216],
       [-0.48359791],
       [-0.40308428],
       [-0.48359791],
       [-0.42992216],
       [-0.51043579],
       [-0.42992216],
       [-0.51043579],
       [-0.53727367],
       [-0.48359791],
       [-0.45676003],
       [-0.51043579],
       [-0.51043579],
       [-0.51043579],
       [-0.34940852],
       [-0

In [ ]:
predicted_y

array([[ 1.7690023 ],
       [ 4.159657  ],
       [ 5.3103685 ],
       [ 0.43682414],
       [ 1.9025725 ],
       [-0.49508578],
       [-0.06380792],
       [ 5.312699  ],
       [11.43319   ],
       [ 1.8413907 ],
       [ 3.1783695 ],
       [ 1.595602  ],
       [ 8.621457  ],
       [-0.96984476],
       [ 0.28122926],
       [-0.41472226],
       [-0.08603241],
       [-0.56457883],
       [-0.22013925],
       [ 3.039267  ],
       [ 9.049262  ],
       [ 2.848444  ],
       [ 6.2099257 ],
       [ 3.2595384 ],
       [ 4.7321177 ],
       [ 3.6679962 ],
       [-0.6302423 ],
       [ 0.14213356],
       [ 0.0887489 ],
       [-0.30074984],
       [ 3.4403462 ],
       [ 6.087864  ],
       [-0.64002776],
       [-0.6871026 ],
       [-0.29875106],
       [ 0.60949624],
       [-0.09720229],
       [-0.81575733],
       [ 1.9033526 ],
       [-0.5167733 ],
       [ 9.115132  ],
       [ 2.1223204 ],
       [ 0.20551424],
       [-0.2090983 ],
       [ 0.16854697],
       [-0

In [ ]:

predicted_y_unscaled = input_scaler.inverse_transform(predicted_y)
y_unseen_unscaled = input_scaler.inverse_transform(y_unseen_scaled)

In [ ]:
predicted_y_unscaled

array([[ 8.89336243e+01],
       [ 1.78011246e+02],
       [ 2.20887650e+02],
       [ 3.92956352e+01],
       [ 9.39105530e+01],
       [ 4.57195377e+00],
       [ 2.06416988e+01],
       [ 2.20974472e+02],
       [ 4.49028717e+02],
       [ 9.16308746e+01],
       [ 1.41447723e+02],
       [ 8.24725952e+01],
       [ 3.44261353e+02],
       [-1.31179304e+01],
       [ 3.34980507e+01],
       [ 7.56635952e+00],
       [ 1.98135967e+01],
       [ 1.98258865e+00],
       [ 1.48166723e+01],
       [ 1.36264664e+02],
       [ 3.60201691e+02],
       [ 1.29154449e+02],
       [ 2.54405838e+02],
       [ 1.44472137e+02],
       [ 1.99341583e+02],
       [ 1.59691589e+02],
       [-4.64081824e-01],
       [ 2.83152370e+01],
       [ 2.63260841e+01],
       [ 1.18130598e+01],
       [ 1.51209183e+02],
       [ 2.49857727e+02],
       [-8.28696311e-01],
       [-2.58274174e+00],
       [ 1.18875360e+01],
       [ 4.57295303e+01],
       [ 1.93973980e+01],
       [-7.37651539e+00],
       [ 9.3

In [ ]:
y_unseen_unscaled

array([[ 44.],
       [ 42.],
       [ 39.],
       [  6.],
       [  6.],
       [  7.],
       [  6.],
       [  4.],
       [  3.],
       [  3.],
       [  3.],
       [  3.],
       [  3.],
       [  5.],
       [  4.],
       [  5.],
       [  4.],
       [  4.],
       [  4.],
       [ 60.],
       [ 90.],
       [ 40.],
       [ 45.],
       [135.],
       [ 80.],
       [184.],
       [ 11.],
       [  6.],
       [  7.],
       [  4.],
       [  7.],
       [  5.],
       [  8.],
       [  5.],
       [  7.],
       [  4.],
       [  7.],
       [  4.],
       [  3.],
       [  5.],
       [  6.],
       [  4.],
       [  4.],
       [  4.],
       [ 10.],
       [  5.],
       [  5.],
       [  2.],
       [  3.],
       [ 87.],
       [ 80.],
       [ 75.]])

In [ ]:
arr = np.concatenate((predicted_y_unscaled, y_unseen_unscaled),axis=1)

In [ ]:
print(arr)

[[ 8.89336243e+01  4.40000000e+01]
 [ 1.78011246e+02  4.20000000e+01]
 [ 2.20887650e+02  3.90000000e+01]
 [ 3.92956352e+01  6.00000000e+00]
 [ 9.39105530e+01  6.00000000e+00]
 [ 4.57195377e+00  7.00000000e+00]
 [ 2.06416988e+01  6.00000000e+00]
 [ 2.20974472e+02  4.00000000e+00]
 [ 4.49028717e+02  3.00000000e+00]
 [ 9.16308746e+01  3.00000000e+00]
 [ 1.41447723e+02  3.00000000e+00]
 [ 8.24725952e+01  3.00000000e+00]
 [ 3.44261353e+02  3.00000000e+00]
 [-1.31179304e+01  5.00000000e+00]
 [ 3.34980507e+01  4.00000000e+00]
 [ 7.56635952e+00  5.00000000e+00]
 [ 1.98135967e+01  4.00000000e+00]
 [ 1.98258865e+00  4.00000000e+00]
 [ 1.48166723e+01  4.00000000e+00]
 [ 1.36264664e+02  6.00000000e+01]
 [ 3.60201691e+02  9.00000000e+01]
 [ 1.29154449e+02  4.00000000e+01]
 [ 2.54405838e+02  4.50000000e+01]
 [ 1.44472137e+02  1.35000000e+02]
 [ 1.99341583e+02  8.00000000e+01]
 [ 1.59691589e+02  1.84000000e+02]
 [-4.64081824e-01  1.10000000e+01]
 [ 2.83152370e+01  6.00000000e+00]
 [ 2.63260841e+01  7

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
series1 = pd.Series(predicted_y_unscaled)

ValueError: ignored

In [ ]:
#add 'blocks' array as new column in DataFrame
df_test['PredictedY'] = arr.tolist()

In [ ]:
df_test.head(52)

LABfeed Sulphur  T1 : FRESH FEED FLOW TO KHDS KGPerHour  \
0              1300                             81175.10664   
1              1500                             79355.24688   
2              1500                             80839.47292   
3              1400                             80525.99544   
4              1400                             92100.39453   
5              1400                             92347.00875   
6              1600                             92390.11732   
7              1600                             91784.62885   
8              1600                             92101.06094   
9              1500                             92318.30391   
10             1500                             92033.90365   
11             1400                             92512.53357   
12             1300                             92351.14547   
13             1600                             92784.83563   
14             1500                             91875.94107   
15             1900                             90077.43161   
16             2200                             92604.82411   
17             1700                             90927.97539   
18             1900                             90162.24609   
19             1900                             74726.91146   
20             2000                             80970.98698   
21             1500                             75612.05703   
22             1500                             75000.02234   
23             1600                             70093.06997   
24             1600                             80652.55273   
25             2600                             70598.77969   
26             2300                             79796.45313   
27             2100                             89573.03758   
28             2200                             89999.33719   
29             2300                             90355.31328   
30             2500                             91735.11120   
31             2400                             91390.05375   
32             2300                             92341.43820   
33             2300                             91925.80313   
34             2300                             91249.33581   
35             2600                             86349.55719   
36             2500                             91765.57885   
37             2100                             82044.83648   
38             2200                             82039.73313   
39             2200                             92076.66961   
40             2100                             92163.17737   
41             2100                             91496.26797   
42             2200                             91732.92813   
43             2100                             91419.13055   
44             2800                             91858.49776   
45             2400                             92040.24135   
46             2400                             91493.83453   
47             1700                             91084.62026   
48             1500                             91909.26891   
49             2000                             81122.00469   
50             2000                             79850.37044   
51             1900                             80887.92000   

    T1 : Reactor inlet temperature DEGC  \
0                            269.946063   
1                            270.202305   
2                            270.058107   
3                            288.456482   
4                            291.011409   
5                            290.059319   
6                            293.125296   
7                            294.031160   
8                            294.016579   
9                            294.009970   
10                           293.986267   
11                           291.534757   
12                           292.931919   
13                           291.957052   
14                 

In [ ]:
from keras.models import Sequential
from tensorflow.keras.optimizers import Adam# define model
from keras.optimizer_v2.adam import Adam

In [ ]:
## KHDS_KerasTuner_Optimized_NN_2 ==> from here we picked the best model with best r2 so far!!
##PASS 10: ==> lETS replicate pass 2 --> PASS 2: This has best result so far on both test & train = R2. 
from keras.activations import linear
batch_size = 10
epochs = 200

# build the model
#input_dim = X_train_scaled.shape[1]

lr_model = Sequential()
lr_model.add(Dense(32, input_shape=(33,), activation=tf.nn.relu, kernel_initializer='he_uniform')) 
lr_model.add(Dropout(0.1))
lr_model.add(Dense(64, kernel_initializer='he_uniform', activation=tf.nn.relu)) #from 64 to 128
lr_model.add(Dropout(0.1))
lr_model.add(Dense(64, kernel_initializer='he_uniform', activation=tf.nn.relu)) #from 64 to 128
lr_model.add(Dropout(0.1))
#lr_model.add(Dense(256, kernel_initializer='he_uniform', activation=tf.nn.relu)) #from 64 to 128
#lr_model.add(Dropout(0.1))
lr_model.add(Dense(1, kernel_initializer='he_uniform',activation='linear'))

# compile the model
#lr_model.compile(loss='mean_squared_error',optimizer=sgd)
#lr_model.compile(loss='mse', optimizer=tf.keras.optimizers.Adam(0.0001), metrics=['mse', 'mae', 'mape'])
lr_model.compile(loss='mse', optimizer=tf.keras.optimizers.Adam(0.001), metrics=['mse'])

# Pass several parameters to 'EarlyStopping' function and assigns it to 'earlystopper'
earlystopper = EarlyStopping(monitor='mse', min_delta=0, patience=10, verbose=1, mode='auto') #patience reduced to 10 from 15

In [ ]:
#X_train, X_test, y_train, y_test = train_test_split(X_scaled, y_scaled, test_size=0.10, random_state=2021)

In [ ]:
X_train.shape, y_train.shape

((639, 33), (639, 1))

In [ ]:
X_test.shape, y_test.shape

((72, 33), (72, 1))

In [ ]:
X_scaled.shape, y_scaled.shape

((711, 33), (711, 1))

In [ ]:
#y_test


In [ ]:
history_5_1 = lr_model.fit(X_train, y_train, batch_size=30, epochs = 500, validation_data=(X_test, y_test),shuffle = True, verbose = 0, callbacks = [earlystopper])

Epoch 39: early stopping


In [ ]:
#history_5_2 = lr_model.fit(X, y, batch_size=30, epochs = 500, shuffle = True, verbose = 0, callbacks = [earlystopper])

Epoch 78: early stopping


In [ ]:
# Runs model with its current weights on the training and testing data
y_train_pred = lr_model.predict(X_train)
y_test_pred = lr_model.predict(X_test)

In [ ]:
import numpy as np
from sklearn.metrics import mean_absolute_percentage_error
# Calculates and prints r2 score of training and testing data
print("The R2 score on the Train set is:\t{:0.3f}".format(r2_score(y_train, y_train_pred)))
print("The RMSE score on the Train set is:\t{:0.3f}".format(np.sqrt(mean_squared_error(y_train, y_train_pred))))
print("The MSE score on the Test set is:\t{:0.3f}".format(mean_squared_error(y_train, y_train_pred)))
print("The MSA score on the Test set is:\t{:0.3f}".format(mean_absolute_error(y_train, y_train_pred)))
print("The MAPE score on the Test set is:\t{:0.3f}".format(mean_absolute_percentage_error(y_train, y_train_pred)))

The R2 score on the Train set is:	0.782
The RMSE score on the Train set is:	0.480
The MSE score on the Test set is:	0.231
The MSA score on the Test set is:	0.272
The MAPE score on the Test set is:	1.007


In [ ]:
print("Now scores on TEST data")
#print("The R2 score on the Test set is:\t{:0.3f}".format(r2_score(y_test, y_test_pred)))
print("The RMSE score on the Test set is:\t{:0.3f}".format(np.sqrt(mean_squared_error(y_test, y_test_pred))))
print("The MSE score on the Test set is:\t{:0.3f}".format(mean_squared_error(y_test, y_test_pred)))
print("The MSA score on the Test set is:\t{:0.3f}".format(mean_absolute_error(y_test, y_test_pred)))
print("The MAPE score on the Test set is:\t{:0.3f}".format(mean_absolute_percentage_error(y_test, y_test_pred)))

Now scores on TEST data


ValueError: ignored

In [ ]:
#!pip install h5py

In [ ]:
from keras.models import model_from_json
from keras.models import load_model

In [ ]:
lr_model.save('my_keras_model.h5')  # creates a HDF5 file 'my_model.h5'

In [ ]:
model = load_model('my_keras_model.h5')

In [ ]:
predicted_y_NN = model.predict(X_unseen_scaled)

In [ ]:
from sklearn import metrics
score = np.sqrt(metrics.mean_squared_error(predicted_y_NN,y_unseen_scaled))
print(f"Before save score (RMSE): {score}")
# Measure MSE error.  
score = metrics.mean_squared_error(predicted_y_NN,y_unseen_scaled)
print("Final score (MSE): {}".format(score))
score1 = metrics.mean_absolute_error(predicted_y_NN,y_unseen_scaled)
print("Final score (MAE): {}".format(score1))
score2 = metrics.r2_score(predicted_y_NN,y_unseen_scaled)
print("Final score (R2): {}".format(score2))
score3 = metrics.mean_absolute_percentage_error(predicted_y_NN,y_unseen_scaled)
print("Final score (MAPE): {}".format(score3))
#score4 = metrics.mean_squared_log_error(predicted_y_scaled,testy_scaled)
#print("Final score (MSLE): {}".format(score4))
#score5 = metrics.log_loss(predicted_y_scaled,testy_scaled)

Before save score (RMSE): 0.9739276638962935
Final score (MSE): 0.9485350945024916
Final score (MAE): 0.7053202218871792
Final score (R2): -110.35914857588583
Final score (MAPE): 25.635091856947692


In [ ]:
predicted_y_unscaled = input_scaler.inverse_transform(predicted_y_NN)
y_unseen_unscaled = input_scaler.inverse_transform(y_unseen_scaled)

In [ ]:
arr_NN = np.concatenate((predicted_y_unscaled, y_unseen_unscaled),axis=1)

In [ ]:
#add 'blocks' array as new column in DataFrame
df_test['PredictedY'] = arr_NN.tolist()

In [ ]:
df_test.head(52)

LABfeed Sulphur  T1 : FRESH FEED FLOW TO KHDS KGPerHour  \
0              1300                             81175.10664   
1              1500                             79355.24688   
2              1500                             80839.47292   
3              1400                             80525.99544   
4              1400                             92100.39453   
5              1400                             92347.00875   
6              1600                             92390.11732   
7              1600                             91784.62885   
8              1600                             92101.06094   
9              1500                             92318.30391   
10             1500                             92033.90365   
11             1400                             92512.53357   
12             1300                             92351.14547   
13             1600                             92784.83563   
14             1500                             91875.94107   
15             1900                             90077.43161   
16             2200                             92604.82411   
17             1700                             90927.97539   
18             1900                             90162.24609   
19             1900                             74726.91146   
20             2000                             80970.98698   
21             1500                             75612.05703   
22             1500                             75000.02234   
23             1600                             70093.06997   
24             1600                             80652.55273   
25             2600                             70598.77969   
26             2300                             79796.45313   
27             2100                             89573.03758   
28             2200                             89999.33719   
29             2300                             90355.31328   
30             2500                             91735.11120   
31             2400                             91390.05375   
32             2300                             92341.43820   
33             2300                             91925.80313   
34             2300                             91249.33581   
35             2600                             86349.55719   
36             2500                             91765.57885   
37             2100                             82044.83648   
38             2200                             82039.73313   
39             2200                             92076.66961   
40             2100                             92163.17737   
41             2100                             91496.26797   
42             2200                             91732.92813   
43             2100                             91419.13055   
44             2800                             91858.49776   
45             2400                             92040.24135   
46             2400                             91493.83453   
47             1700                             91084.62026   
48             1500                             91909.26891   
49             2000                             81122.00469   
50             2000                             79850.37044   
51             1900                             80887.92000   

    T1 : Reactor inlet temperature DEGC  \
0                            269.946063   
1                            270.202305   
2                            270.058107   
3                            288.456482   
4                            291.011409   
5                            290.059319   
6                            293.125296   
7                            294.031160   
8                            294.016579   
9                            294.009970   
10                           293.986267   
11                           291.534757   
12                           292.931919   
13                           291.957052   
14                 

In [ ]:
#target_col = "Product Sulphur"
#X_unseen = df_test.loc[:, df_test.columns != target_col]
#y_unseen = df_test.loc[:, target_col]

In [ ]:
#history_5_2 = lr_model.fit(X_unseen, y_unseen, batch_size=30, epochs = 500, shuffle = True, verbose = 0, callbacks = [earlystopper])

Epoch 44: early stopping


In [ ]:
# Runs model with its current weights on the training and testing data
#y_train_pred = lr_model.predict(X_train)
y_test_pred = lr_model.predict(X_unseen)

In [ ]:
from sklearn import metrics
score = np.sqrt(metrics.mean_squared_error(predicted_y_NN,y_unseen_scaled))
print(f"Before save score (RMSE): {score}")
# Measure MSE error.  
score = metrics.mean_squared_error(predicted_y_NN,y_unseen_scaled)
print("Final score (MSE): {}".format(score))
score1 = metrics.mean_absolute_error(predicted_y_NN,y_unseen_scaled)
print("Final score (MAE): {}".format(score1))
score2 = metrics.r2_score(predicted_y_NN,y_unseen_scaled)
print("Final score (R2): {}".format(score2))
score3 = metrics.mean_absolute_percentage_error(predicted_y_NN,y_unseen_scaled)
print("Final score (MAPE): {}".format(score3))
#score4 = metrics.mean_squared_log_error(predicted_y_scaled,testy_scaled)
#print("Final score (MSLE): {}".format(score4))
#score5 = metrics.log_loss(predicted_y_scaled,testy_scaled)

Before save score (RMSE): 2.764894541203146
Final score (MSE): 7.644641823974954
Final score (MAE): 1.4691307036457955
Final score (R2): 0.08199124250458523
Final score (MAPE): 6.480537202562965


In [ ]:
pred_y_NN1 = model.predict(X_unseen_scaled)

In [ ]:
from sklearn import metrics
score = np.sqrt(metrics.mean_squared_error(pred_y_NN1,y_unseen_scaled))
print(f"Before save score (RMSE): {score}")
# Measure MSE error.  
score = metrics.mean_squared_error(pred_y_NN1,y_unseen_scaled)
print("Final score (MSE): {}".format(score))
score1 = metrics.mean_absolute_error(pred_y_NN1,y_unseen_scaled)
print("Final score (MAE): {}".format(score1))
score2 = metrics.r2_score(pred_y_NN1,y_unseen_scaled)
print("Final score (R2): {}".format(score2))
score3 = metrics.mean_absolute_percentage_error(pred_y_NN1,y_unseen_scaled)
print("Final score (MAPE): {}".format(score3))
#score4 = metrics.mean_squared_log_error(predicted_y_scaled,testy_scaled)
#print("Final score (MSLE): {}".format(score4))
#score5 = metrics.log_loss(predicted_y_scaled,testy_scaled)

Before save score (RMSE): 2.764894541203146
Final score (MSE): 7.644641823974954
Final score (MAE): 1.4691307036457955
Final score (R2): 0.08199124250458523
Final score (MAPE): 6.480537202562965
